libraries/standard/mqtt/src/mqtt_lightweight.c

```diff
@@ -192,6 +252,515 @@ static int32_t recvExact( MQTTTransportRecvFunc_t recvFunc,
+        {
+            "Connection accepted.",                               /* 0 */
+            "Connection refused: unacceptable protocol version.", /* 1 */
+            "Connection refused: identifier rejected.",           /* 2 */
+            "Connection refused: server unavailable",             /* 3 */
+            "Connection refused: bad user name or password.",     /* 4 */
+            "Connection refused: not authorized."                 /* 5 */
+        };
+        IotLogErrorWithArgs( "%s", pConnackResponses[ responseCode ] );
+    #endif
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t _deserializeConnack( const MQTTPacketInfo_t * const pConnack,
+                                         bool * const pSessionPresent )
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    assert( pConnack != NULL );
+    assert( pSessionPresent != NULL );
+    const uint8_t * pRemainingData = pConnack->pRemainingData;
+
+    /* According to MQTT 3.1.1, the second byte of CONNACK must specify a
+     * "Remaining length" of 2. */
+    if( pConnack->remainingLength != MQTT_PACKET_CONNACK_REMAINING_LENGTH )
+    {
+        IotLogErrorWithArgs( "CONNACK does not have remaining length of %d.",
+                             MQTT_PACKET_CONNACK_REMAINING_LENGTH );
+
+        status = MQTTBadResponse;
+    }
+
+    /* Check the reserved bits in CONNACK. The high 7 bits of the second byte
+     * in CONNACK must be 0. */
+    else if( ( pRemainingData[ 0 ] | 0x01U ) != 0x01U )
+    {
+        IotLogError( "Reserved bits in CONNACK incorrect." );
+
+        status = MQTTBadResponse;
+    }
+    else
+    {
+        /* Determine if the "Session Present" bit is set. This is the lowest bit of
+         * the second byte in CONNACK. */
+        if( ( pRemainingData[ 0 ] & MQTT_PACKET_CONNACK_SESSION_PRESENT_MASK )
+            == MQTT_PACKET_CONNACK_SESSION_PRESENT_MASK )
+        {
+            IotLogWarn( "CONNACK session present bit set." );
+            *pSessionPresent = true;
+
+            /* MQTT 3.1.1 specifies that the fourth byte in CONNACK must be 0 if the
+             * "Session Present" bit is set. */
+            if( pRemainingData[ 1 ] != 0U )
+            {
+                status = MQTTBadResponse;
+            }
+        }
+        else
+        {
+            IotLogInfo( "CONNACK session present bit not set." );
+        }
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* In MQTT 3.1.1, only values 0 through 5 are valid CONNACK response codes. */
+        if( pRemainingData[ 1 ] > 5U )
+        {
+            IotLogErrorWithArgs( "CONNACK response %hhu is not valid.",
+                                 pRemainingData[ 1 ] );
+
+            status = MQTTBadResponse;
+        }
+        else
+        {
+            /* Print the appropriate message for the CONNACK response code if logs are
+             * enabled. */
+            #if LIBRARY_LOG_LEVEL > IOT_LOG_NONE
```

Comment:
I think #if is not needed (since already inside the function )

Reply:
Yep, I had forgotten to remove one of the macro guards